src/wp-includes/block-template-utils.php

```diff
@@ -1432,39 +1432,45 @@ function get_template_hierarchy( $slug, $is_custom = false, $template_prefix = '
 	$template_hierarchy[] = 'index';
 	return $template_hierarchy;
 }
+
 /**
  * Inject ignoredHookedBlocks metadata attributes into a template or template part.
  *
- * Given a `wp_template` or `wp_template_part` post object, locate all blocks that have
+ * Given an object that represents a `wp_template` or `wp_template_part` post object
+ * prepared for inserting or updating the database, locate all blocks that have
  * hooked blocks, and inject a `metadata.ignoredHookedBlocks` attribute into the anchor
  * blocks to reflect the latter.
  *
- * @param WP_Post $post A post object with post type set to `wp_template` or `wp_template_part`.
- * @return WP_Post The updated post object.
+ * @param stdClass        $post    An object representing a template or template part
+ *                                 prepared for inserting or updating the database.
+ * @param WP_REST_Request $request Request object.
+ * @return stdClass The updated object representing a template or template part.
  */
-function inject_ignored_hooked_blocks_metadata_attributes( $post ) {
+function inject_ignored_hooked_blocks_metadata_attributes( $post, $request ) {
+	$filter_name = current_filter();
+	if ( ! str_starts_with( $filter_name, 'rest_pre_insert_' ) ) {
+		return $post;
+	}
+	$post_type = str_replace( 'rest_pre_insert_', '', $filter_name );
+
 	$hooked_blocks = get_hooked_blocks();
 	if ( empty( $hooked_blocks ) && ! has_filter( 'hooked_block_types' ) ) {
-		return;
+		return $post;
 	}
 
 	// At this point, the post has already been created.
 	// We need to build the corresponding `WP_Block_Template` object as context argument for the visitor.
 	// To that end, we need to suppress hooked blocks from getting inserted into the template.
 	add_filter( 'hooked_block_types', '__return_empty_array', 99999, 0 );
-	$template = _build_block_template_result_from_post( $post );
+	$template = $request['id'] ? get_block_template( $request['id'], $post_type ) : null;
```

Comment:
We can have two different use cases where the `id` might be missing:
1. The user created a completely new template in the UI.
2. An existing template bundled with a theme not previously stored in the database was updated.

How could we create that `$template` object so the context is properly passed to the logic that updates the serialized blocks?

Reply:
>The user created a completely new template in the UI.

@gziolo have you tested this scenario with this PR? I've just done so myself and wondering if I've missed something. If a user creates a Custom Template they get two routes:

1. The user can start with a blank template. In which case there are no anchor blocks present.
2. The user can select a "Pattern" or starter template. In my testing I chose this route.

As you can see in the screenshots the starter templates come with hooked blocks already attached. In my case below in the screenshot I've hooked the `core/loginout` block after the `core/post-featured-image` block.

![Screenshot 2024-03-06 at 08 24 02](https://github.com/WordPress/wordpress-develop/assets/8639742/240210d5-99db-493b-aa60-46dc1434ae6e)

Upon saving this template, it does indeed have an ID which (I assume) was created the second you chose to "_Add New Template > Custom Template_" so when this is saved the algorithm runs as expected inserting the `ignoredHookedBlocks` meta data on it.

---

>An existing template bundled with a theme not previously stored in the database was updated.

I'm going to test this scenario as well but this is the case with all theme templates unless I'm misunderstanding your concern? The function `get_block_template()` fetches the template object if it exists in the database, or gets it from the filesystem. So it's not exclusive to just database templates.

Reply:
>I'm going to test this scenario as well but this is the case with all theme templates unless I'm misunderstanding your concern? The function `get_block_template()` fetches the template object if it exists in the database, or gets it from the filesystem. So it's not exclusive to just database templates.

Adding to this: Maybe the source of confusion is the nature of `$request['id']`. This is a template (or template part) ID, i.e. not a numeric post object ID from the database, but a string composed from the theme and template names, such as `twentytwentyfour//single` or `twentytwentythree//header`. These IDs do work for templates that are fetched straight from the corresponding theme file.

Reply:
My intention was to ensure it wasn’t overlooked as there are no unit tests included in the PR to validate that.

Reply:
I could confirm in my testing that `$request['id']` is properly set for templates sourced from HTML files that are about to get saved to the database 👍🏻